### PR TITLE
Enable e2ee across all services and save crypto material in the database

### LINKED ISSUE #270: olm?
Hello looking for matrix go libs to create a bot but needs to support OLM etc since all channels are encrypted now a days.(most of them at least in our homeserver) does go-neb supports accepting keys by default or importing manually ? will it ever support it?

Thanks.

### PR DESCRIPTION
**Pull request by nikofil**
* Move some client logic to a new BotClient type
* Persist crypto data to a database if provided (relies on https://github.com/tulir/mautrix-go/pull/10)
* Enable e2ee for all neb services
* Sync the full state of joined rooms when starting up

Please note you also have to provide a device ID in the config along with the access token for this to work.

Closes #270

### CODE REVIEW

**Comment by Kegsay** [comment]
Please fix Travis :)

**Comment by nikofil** [comment]
Ah sorry, I missed the Travis build somehow and I didn't see the tests were hanging. Fixed now.

File: clients/clients.go [hunk 1]
```diff
@@ -341,47 +335,106 @@ func (c *Clients) onRoomMemberEvent(client *mautrix.Client, event *mevt.Event) {
...
+func (c *Clients) initClient(botClient *BotClient) error {
+	config := botClient.config
 	client, err := mautrix.NewClient(config.HomeserverURL, config.UserID, config.AccessToken)
 	if err != nil {
-		return nil, err
+		return err
 	}
+
 	client.Client = c.httpClient
+	client.DeviceID = config.DeviceID
```
Comment (Kegsay) [inline clients/clients.go:346]: What happens if no device ID is provided for the client? We should support this (and just not have E2E work).

Comment (nikofil) [reply clients/clients.go:346]: That's what happens - the device ID is only used for e2ee operations. I added a warning to let the user know they didn't specify a device ID, other than that unencrypted rooms should work normally.

**Comment by electricOzone** [comment]
I just pulled down the latest docker image, and it appears this is still broken in unencrypted rooms:
```
time="2020-06-29T19:28:50Z" level=info msg="Executing command" command="[giphy]" room_id= user_id="@electricozone:services.pyrahex.com"
time="2020-06-29T19:28:50Z" level=info msg="Searching giphy for docker"
time="2020-06-29T19:28:51Z" level=info msg="Failed to send command response" content="{m.image season-6-the-simpsons-6x18-xT5LMsC6iydeziO6nC    mxc://services.pyrahex.com/lRrmlobzOApbBbgwQOoNygqo 0xc0006e0050 <nil> <nil> <nil> false}" error="msg=Failed to PUT JSON to /_matrix/client/r0/rooms/send/m.room.message/mautrix-go_1593458931039470062_2 code=400 wrapped=M_UNRECOGNIZED: Unrecognized request" room_id= user_id="@electricozone:services.pyrahex.com"
```

**Comment by nikofil** [comment]
Are you sure you have the latest img? `roomid=<blank>` looks to be the error that was fixed in this PR. I just tried the last docker img myself and it worked:
```
INFO[0324] Executing command                             command="[giphy]" room_id="!EzqLMSHOPMnQEDOuVU:localhost:8008" user_id="@admin:localhost:8008"
INFO[0324] Searching giphy for hi
```

**Comment by electricOzone** [comment]
Ah, that's my bad. I neglected to do a docker pull after switching back to matrixdotorg/go-neb:latest. Thanks for all the good work everyone!

**Comment by Kegsay** [comment]
Thanks for acting as the canary in the coalmine as it were @electricOzone - hopefully we can address #326 to make this less likely to happen again.

**Pull request merged**